### PR TITLE
Adds Lightning Holoparasites

### PR DESCRIPTION
**Pull request by ChangelingRain**
:cl: Joan
rscadd: Adds lightning holoparasites, which have medium damage resist, a weak attack, have a lightning chain to their summoner, and apply lightning chains when attacking targets.
rscadd: Lightning chains shock everyone nearby, doing low burn damage. This is excluding the parasite and summoner; chains will shock enemies they're attached to.
tweak: Holoparasites can now smash tables and lockers.
/:cl:

shocking
Lightning chain notes;
Chains shock everyone except the summoner and parasite
Chains shock anyone even vaguely near them
Chains to enemies will shock them, too
Chains applied to enemies last for 7 seconds or until the enemy moves out of range
Attacking the same enemy doesn't reset the chain duration on that enemy
You can have up to 3 chains to enemies at once; new targets will remove a random existing chain
Chains do _not_ stun
Chains do 5 burn damage, ignoring armor, when they shock something
If chain damage areas overlap, each chain will shock enemies individually, meaning more chains == more damage

### CODE REVIEW

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 1]
```diff
@@ -435,6 +437,95 @@
...
+		if(enemychain)
+			qdel(enemychain)
+			enemychain = null
+			enemy = null
+
+/mob/living/simple_animal/hostile/guardian/beam/proc/chainshock(var/datum/beam/B)
+	var/list/turfs = list()
+	for(var/E in B.elements)
+		var/obj/effect/ebeam/chainpart = E
+		turfs |= get_turf(chainpart)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: I told you to use ```get_turf_pixel()```, ```get_turf()``` will return the starting turf of the beam ```B.elements.len times```, which is wrong on two levels.

```get_turf_pixel()``` on the other hand WILL get you the turfs where the beam looks like it is.

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 2]
```diff
@@ -435,6 +437,95 @@
...
+	for(var/turf in getline(B.origin, B.target))
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: changing the above, this becomes overkill.

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 3]
```diff
@@ -435,6 +437,95 @@
...
+		var/turf/T = turf
+		turfs |= T
+	for(var/turf in turfs)
+		var/turf/T = turf
+		if(T != get_turf(B.origin) && T != get_turf(B.target))
+			for(var/turf/TU in circlerange(T, 1))
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: this just adds the 8 turfs around TU to the list right? you can do that in the first loop.

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 1 above)
Comment (ChangelingRain) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: get_turf_pixel() gives me a runtime;

related code segment(other code is the same);
```
	for(var/E in B.elements)
		var/obj/effect/ebeam/chainpart = E
		var/turf/T = get_turf_pixel(chainpart)
		turfs |= T
		if(T != get_turf(B.origin) && T != get_turf(B.target))
			for(var/turf/TU in circlerange(T, 1))
				turfs |= TU
```

**Comment by Bawhoppen** [comment]
while you're in their increase their tc cost :panda_face:

File: code/modules/mob/living/simple_animal/guardian/guardian.dm (hunk 1 above)
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: oh gosh that's hella bad.
The offending line is ```get_turf()```, ```get_turf()``` will only return null if given null.

I think your list has nulls in it!

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 4]
```diff
@@ -435,6 +437,95 @@
...
+	if(..())
+		if(summonerchain)
+			qdel(summonerchain)
+			summonerchain = null
+		if(enemychain)
+			qdel(enemychain)
+			enemychain = null
+			enemy = null
+
+/mob/living/simple_animal/hostile/guardian/beam/proc/chainshock(var/datum/beam/B)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: no ```var/``` needed, just ```datum/beam/B```

File: code/modules/mob/living/simple_animal/guardian/guardian.dm [hunk 5]
```diff
@@ -435,6 +437,92 @@
...
+				qdel(enemychain)
+				enemychain = null
+			enemy = target
+			enemychain = Beam(enemy,"lightning[rand(1,12)]",'icons/effects/effects.dmi',70, 8,/obj/effect/ebeam/chain)
+
+/mob/living/simple_animal/hostile/guardian/beam/Life()
+	..()
+	if(summoner && summonerchain && !qdeleted(summonerchain))
+		chainshock(summonerchain)
+	else if(summoner)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/guardian/guardian.dm]: but this and the check above under one ```if(summoner)```
that's 1 less scope traversal and one less statement evaluated.

**Comment by Incoming5643** [comment]
if the stun is longer than the cooldown between electropunches this seems OP as hell

**Comment by KazeEspada** [comment]
CHAINS DO NOT STUN

**Comment by Bawhoppen** [comment]
what a bad bad world of balance we have come to when someone immediately assumes stunning when there is a new attack

**Comment by PKPenguin321** [comment]
well i mean it is a lightning attack that looks exactly like the wizard's lightning which does stun, so can you blame them?

**Comment by ChangelingRain** [comment]
It's not exactly an unlikely assumption, but it does say "does not stun" right in the main post

**Comment by Buggy123** [comment]
:+1: Holo's are generally horrible against groups, except maybe chaos parasites, so something that is MORE effective against groups is a nice change of pace.

**Pull request merged**